controllers/notebook/notebook_controller.go

```diff
@@ -0,0 +1,408 @@
+// +kubebuilder:rbac:groups="",resources=services,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups="",resources=services/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups="networking.k8s.io",resources=ingresses,verbs=get;list;watch;create;update;patch;delete
+
+func (r *NotebookReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	_ = r.Log.WithValues("notebook", req.NamespacedName)
+
+	notebook := &notebookv1alpha1.Notebook{}
+	err := r.Get(context.Background(), req.NamespacedName, notebook)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			r.Log.Info("notebook doesn't exist", "name", req.String())
+			return reconcile.Result{}, nil
+		}
+		r.Log.Error(err, "fail to get notebook", "name", req.String())
+		return reconcile.Result{}, err
+	}
+
+	if notebook.Status.Condition == notebookv1alpha1.NotebookTerminated {
+		r.Log.Info("notebook terminated", "name", req.String())
+		return reconcile.Result{}, nil
+	}
+
+	// 1. create the notebook pod if not existing
+	pod := &corev1.Pod{}
+	nbPodName := types.NamespacedName{Namespace: notebook.Namespace, Name: nbNamePrefix(notebook.Name)}
+	err = r.Get(context.Background(), nbPodName, pod)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			r.Log.Info("create notebook pod " + nbPodName.String())
+			pod.ObjectMeta = *notebook.Spec.Template.ObjectMeta.DeepCopy()
+			pod.Name = nbPodName.Name
+			pod.Namespace = nbPodName.Namespace
+			if pod.Labels == nil {
+				pod.Labels = make(map[string]string, 1)
+			}
+			pod.Labels[notebookv1alpha1.NotebookNameLabel] = notebook.Name
+			pod.Spec = notebook.Spec.Template.Spec
+
+			addNotebookDefaultPort(&pod.Spec)
+
+			log.Infof("pod spec: %v", pod.Spec.Containers[0])
+			err := setJuypterLabEnv(pod)
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+
+			// add notebook as pod's owner reference, so that the pod is deleted when notebook is deleted
+			addOwnerReferenceIfMissing(&pod.ObjectMeta, notebook)
+
+			// create the notebook pod
+			err = r.Create(context.Background(), pod)
+			if err != nil {
+				r.Log.Error(err, "failed to create pod "+nbPodName.String())
+				return reconcile.Result{Requeue: true}, err
+			}
+			r.recorder.Eventf(notebook, core.EventTypeNormal, "SuccessfulCreatedPod", "Created pod %v", pod.Name)
+			notebook.Status.Condition = notebookv1alpha1.NotebookCreated
+			notebook.Status.Message = "created notebook pod " + pod.Name
+			notebook.Status.LastTransitionTime = metav1.Now()
+			err = r.Status().Update(context.Background(), notebook)
+			if err != nil {
+				r.Log.Error(err, "failed to update notebook status", "notebook", notebook.Name)
+				return ctrl.Result{}, err
+			}
+		} else {
+			// for other errors
+			return reconcile.Result{}, err
+		}
+	}
+
+	// check pod status and update notebook status
+	if pod.Status.Phase == corev1.PodPending {
+		if notebook.Status.Condition == "" {
+			notebook.Status.Condition = notebookv1alpha1.NotebookCreated
```

Comment:
update notebook status conditions can be encapsulated into an independent function

Reply:
done